[mlir][LLVM] Improve lowering of `llvm.byval` function arguments

dcaballe opened the pull request:
When a function argument is annotated with the `llvm.byval` attribute, [LLVM expects](https://llvm.org/docs/LangRef.html#parameter-attributes) the function argument type to be an `llvm.ptr`. For example:

```
func.func (%args0 : llvm.ptr {llvm.byval = !llvm.struct<(i32)>} {
  ...
}
```

Unfortunately, this makes the type conversion context-dependent, which is something that the type conversion infrastructure (i.e., `LLVMTypeConverter` in this particular case) doesn't support. For example, we may want to convert `MyType` to `llvm.struct<(i32)>` in general, but to an `llvm.ptr` type only when it's a function argument passed by value.

To fix this problem, this PR changes the FuncToLLVM conversion logic to generate an `llvm.ptr` when the function argument has a `llvm.byval` attribute. An `llvm.load` is inserted into the function to retrieve the value expected by the argument users.

Dinistro commented:
Will this not lead to inconsistencies in the types?
Given your `MyType` example, this could result in IR where the argument is now an `llvm.ptr`, while all its usages expect an `llvm.struct<(i32)>`. Is this supposed to be fixed by some additional meterializations, or how would one deal with such cases?

dcaballe commented:
Good point! I see how that is automatically handled for my case... but that's may not be the cases in general.
We can address this by introducing a load (e.g., `%0 = llvm.load %arg0 : !llvm.ptr -> !llvm.struct<(...)>`) and replacing all the uses of the argument with it. IIRC, there was an easy way to do that for function arguments... Let me look into it. Thanks!

dcaballe committed:
[mlir][LLVM] Improve lowering of llvm.byval function arguments

When a function argument is annotated with the `llvm.byval` attribute,
[LLVM expects] (https://llvm.org/docs/LangRef.html#parameter-attributes)
the function argument type to be an `llvm.ptr`. For example:

```
func.func (%args0 : llvm.ptr {llvm.byval = !llvm.struct<(i32)>} {
  ...
}
```

Unfortunately, this makes the type conversion context-dependent, which is
something that the type conversion infrastructure (i.e., `LLVMTypeConverter`
in this particular case) doesn't support. For example, we may want to convert
`MyType` to `llvm.struct<(i32)>` in general, but to an `llvm.ptr` type only
when it's a function argument passed by value.

To fix this problem, this PR changes the FuncToLLVM conversion logic to always
generate an `llvm.ptr` when the function argument has a `llvm.byval` attribute.
An `llvm.load` is inserted into the function to retrieve the value expected by
the argument users.

joker-eph commented:
Isn't there a problem with the way `llvm.byval` seems modeled here: why is this something that annotating before the lowering instead of a property of the type you lower?

dcaballe commented:
> Isn't there a problem with the way `llvm.byval` seems modeled here: why is this something that annotating before the lowering instead of a property of the type you lower? 

Good point! There are a few aspects related to this: 

* There seems to be multiple users of this “pre-lowering annotation” approach, including FIR and probably some downstreams, which goes beyond the `llvm.byval` case. We can see how a few attributes are processed [here](https://github.com/llvm/llvm-project/blob/main/mlir/lib/Conversion/FuncToLLVM/FuncToLLVM.cpp#L353-L369) and they are not inserted by the LLVM lowering itself. In some cases, this is probably an attempt at reusing those attributes in other dialects before LLVM. 

* This `llvm.byval` mechanism seems very specific to LLVM. A dialect that is eventually lowered to LLVM (or any other egress dialect) may model a pass-by-value of a non-ptr type by just passing an SSA value of that type. Unfortunately, if that type is lowered to an aggregate type in LLVM, the pass-by-value will have to go through the `llvm.byval` mechanism, which requires passing an `llvm.ptr` instead of the actual aggregate type. To me, this looks like an LLVM ABI requirement more than a property of the type lowered to LLVM. 

* The `llvm.byval` mechanism can also be used for non-aggregate types (not sure why we would use it instead of just passing the non-aggregate type directly). This makes the previous bullet also applicable to non-aggregate types. 

I’m planning to follow up on this and automatically introduce the `llvm.byval` attributes for aggregate type arguments. However, that won’t solve the problem for non-aggregate types and other LLVM attributes that are currently introduced before the lowering.

joker-eph commented:
Taking another angle, the property to keep is that the type converter is always 1:1, that is for a given type before conversion you always get the same converted type.
We could do that and handle the byval conversion separately (outside of the type converter) directly in the signature conversion logic targeting llvm.func? (the logic would have to add the right `load` to the converted type as well).

dcaballe commented:
> We could do that and handle the byval conversion separately (outside of the type converter) directly in the signature conversion logic targeting llvm.func? (the logic would have to add the right load to the converted type as well).

It sounds reasonable. I was hoping to avoid creating two function signatures but let me give it a try...

joker-eph commented:
> I was hoping to avoid creating two function signatures but let me give it a try...

I'm not sure what you're referring to? To be clear: in my mind there should be no change to the TypeConverter or the dialect conversion framework in general.

dcaballe commented:
I meant creating two converted function types if we introduced the `llvm.ptr` argument types after [this point](https://github.com/llvm/llvm-project/blob/main/mlir/lib/Conversion/FuncToLLVM/FuncToLLVM.cpp#L307). 

We can also do it **before** that point and create the converted function with the final type. However, it's not ideal either. We would have to update the argument types in the `TypeConverter::SignatureConversion` [returned](https://github.com/llvm/llvm-project/blob/main/mlir/lib/Conversion/FuncToLLVM/FuncToLLVM.cpp#L283) by `converter.convertFunctionSignature`. To some extent, this means the LLVM type converter wouldn't be providing a valid conversion in the first place, which is unexpected. 

We would also have to map the attributes from the original function to the converted function, since the attributes haven’t been [propagated](https://github.com/llvm/llvm-project/blob/main/mlir/lib/Conversion/FuncToLLVM/FuncToLLVM.cpp#L331-L332) to the converted function yet. Finally, we would have to replicate all this functionality in the SPIR-V and GPU lowerings to LLVM...  
  

This is mostly why I encapsulated all this complexity behind the LLVM type converter (I tried a few other things before reaching this point). It makes the type conversion dependent on the argument attributes (which is accurate for the specific case of the function signature, after all) but the LLVM type converter would return a valid converted function type in the first place. 

No strong opinion, though. It’s a matter of trade-offs. Happy to follow the feedback provided but I want to make sure we are all on the same page.

River707 commented:
I agree with Mehdi here. This PR seems to be creating an implicit transformation for a given LLVM attribute, which seems a bit off (I wouldn't necessarily expect the converter to automatically pointer-ize the argument type). It feels like this kind of behavior should be left directly to the llvm func conversion, with additional extensions added to the llvm converter (if it's to continue being used as an extension point) for opting-into/controlling this behavior.

joker-eph commented:
I saw `TypeConverter.h` being modified and thought it was the general dialect conversion, not something specific to LLVM conversion.
It can be reasonable to have dedicated support for LLVM specific things in the LLVMTypeConverter indeed.

dcaballe commented:
> I saw TypeConverter.h being modified and thought it was the general dialect conversion, not something specific to LLVM conversion.

Note that the current changes are to `LLVMTypeConverter` (under `.../LLVMCommon/TypeConverter.h`), not the generic type converter.

Ok, if no other comments, let me try to move the logic to the llvm function conversion so that we can compare both approaches.
Thanks!

dcaballe committed:
Revert "[mlir][LLVM] Improve lowering of llvm.byval function arguments"

This reverts commit eb7cec3712345e5260142886e1fa608ad57ca69b.

dcaballe committed:
Second approach

dcaballe committed:
Fixes and cleanup

River707 commented on mlir/include/mlir/Transforms/DialectConversion.h:
```diff
@@ -75,6 +76,10 @@ class TypeConverter {
 
+    /// Get the converted type for the given argument only if there is a
+    /// one-to-one mapping for it. Otherwise, return std::nullptr.
+    std::optional<Type> getConvertedType(unsigned inputo) const;
```
```suggestion
    std::optional<Type> getConvertedType(unsigned input) const;
```

River707 commented on mlir/lib/Conversion/LLVMCommon/TypeConverter.cpp:
```diff
@@ -302,6 +302,40 @@ Type LLVMTypeConverter::convertFunctionSignature(
+      continue;
+
+    StringRef inAttrName = inAttr->getName().getValue();
+    if (inAttrName != LLVM::LLVMDialect::getByValAttrName() &&
+        inAttrName != LLVM::LLVMDialect::getByRefAttrName())
+      continue;
```
When would this be true? Didn't you already filter by these?

River707 commented on mlir/lib/Conversion/FuncToLLVM/FuncToLLVM.cpp:
```diff
@@ -281,12 +345,20 @@ mlir::convertFuncOpToLLVMFuncOp(FunctionOpInterface funcOp,
-  auto llvmType = converter.convertFunctionSignature(
-      funcTy, varargsAttr && varargsAttr.getValue(),
-      shouldUseBarePtrCallConv(funcOp, &converter), result);
+  auto llvmType =
+      cast_or_null<LLVM::LLVMFunctionType>(converter.convertFunctionSignature(
+          funcTy, varargsAttr && varargsAttr.getValue(),
+          shouldUseBarePtrCallConv(funcOp, &converter), result));
   if (!llvmType)
     return rewriter.notifyMatchFailure(funcOp, "signature conversion failed");
 
+  // Replace the type of `llvm.byval` and `llvm.byref` arguments that were not
+  // converted to an LLVM pointer type.
+  SmallVector<std::optional<NamedAttribute>> byValRefNonPtrAttrs;
+  filterByValRefNonPtrAttrs(funcOp, result, byValRefNonPtrAttrs);
+  llvmType = converter.materializePtrForByValByRefFuncArgs(
+      llvmType, byValRefNonPtrAttrs, result);
```
Feels like the filtering and signature conversion could be in a new `convertFunctionSignature` overload that takes a FuncOp (would encapsulate this a bit more, cleanup the materializePtrForByValByRefFuncArgs logic, and avoid the double type construction), but no super strong feeling.

dcaballe replied:
Yes, that was the motivation for option num 1. I implemented the override and remove a bunch of new methods that are no longer needed. PTAL

dcaballe commented:
Just for reference:
* 1st approach (`LLVMTypeConverter::convertFunctionSignature` returns the final `LLVMFunctionType`): https://github.com/llvm/llvm-project/pull/100028/commits/eb7cec3712345e5260142886e1fa608ad57ca69b
* 2nd approach (The `LLVMFunctionType` is fixed after the main `LLVMTypeConverter::convertFunctionSignature`): https://github.com/llvm/llvm-project/pull/100028/files

River707 reviewed (approved):
The general approach looks good, it would be nice to have the logic a bit cleaner though. Avoiding the secondary pass over the function arguments doing type conversion would probably clean this up a lot. Suppose that may be somewhere in the middle of the two approaches, but I leave things up to you.

dcaballe committed:
Add convertFunctionSignature override

River707 reviewed (approved):
Nice
